atomicapp/providers/lib/kubebase.py

```diff
@@ -0,0 +1,309 @@
+        return plural
+
+    def _contexts(self):
+        '''
+        Parses the contexts and formats it in a name = object way.
+        ex.
+            'foobar': { name: 'foobar', context: 'foo' }
+        '''
+        contexts = {}
+        if "contexts" not in self.config:
+            raise KubeBaseError("No contexts within the .kube/config file")
+        for f in self.config["contexts"]:
+            contexts[f["name"]] = f["context"]
+        return contexts
+
+    def _clusters(self):
+        '''
+        Parses the clusters and formats it in a name = object way.
+        ex.
+            'foobar': { name: 'foobar', cluster: 'foo' }
+        '''
+        clusters = {}
+        if "clusters" not in self.config:
+            raise KubeBaseError("No clusters within the .kube/config file")
+        for f in self.config["clusters"]:
+            clusters[f["name"]] = f["cluster"]
+        return clusters
+
+    def _users(self):
+        '''
+        Parses the users and formats it in a name = object way.
+        ex.
+            'foobar': { name: 'foobar', user: 'foo' }
+        '''
+        users = {}
+        if "users" not in self.config:
+            raise KubeBaseError("No users within the .kube/config file")
+        for f in self.config["users"]:
+            users[f["name"]] = f["user"]
+        return users
+
+    def _connection(self):
+        '''
+        Initializes the required requests session certs / token / authentication
+        in order to communicate with the API
+        '''
+        connection = requests.Session()
+
+        # CA Certificate for TLS verification
+        if "certificate-authority" in self.cluster:
+            connection.verify = self.cert_file(self.cluster["certificate-authority"])
+
+        # Check to see if verification has been disabled, if it has
+        # disable tls-verification
+        if "insecure-skip-tls-verify" in self.cluster:
+            # If it's 'true' within .kube/config
+            if self.cluster["insecure-skip-tls-verify"]:
+                connection.verify = False
+                # Disable the 'InsecureRequestWarning' notifications.
+                # As per: https://github.com/kennethreitz/requests/issues/2214
+                # Instead make a large one-time noticable warning instead
+                requests.packages.urllib3.disable_warnings()
+                logger.warning("CAUTION: TLS verification has been DISABLED")
+            else:
+                logger.debug("Verification will be required for all API calls")
+
+        # If we're using a token, use it, otherwise it's assumed the user uses
+        # client-certificate and client-key
+        if "token" in self.user:
+            connection.headers["Authorization"] = "Bearer %s" % self.user["token"]
+
+        # Lastly, if we have client-certificate and client-key in the .kube/config
+        # we add them to the connection as a cert
+        if ["client-certificate", "client-key"] in self.user.keys():
```

Comment:
We also need support for `client-certificate-data` and `client-key-data` keys.
If specified those options contain "inlined"  base64 encoded certificate/key.

https://github.com/kubernetes/kubernetes/blob/v1.2.2/pkg/client/unversioned/clientcmd/api/types.go#L83,L84